dogstatsd/src/datadog.rs

```diff
@@ -219,12 +234,68 @@ impl DdApi {
             }
         };
 
-        let resp = builder.send().await;
+        let resp = self.send_with_retry(builder, retry_strategy).await;
 
         let elapsed = start.elapsed();
         debug!("Request to {} took {}ms", url, elapsed.as_millis());
-        resp.map_err(|e| ShippingError::Destination(e.status(), format!("Cannot reach {url}")))
+        resp
     }
+
+    async fn send_with_retry(
+        &self,
+        builder: reqwest::RequestBuilder,
+        retry_strategy: RetryStrategy,
+    ) -> Result<Response, ShippingError> {
+        let mut attempts = 0;
+        loop {
+            attempts += 1;
+            match builder.try_clone() {
+                Some(builder) => {
+                    let resp = builder.send().await;
+                    match resp {
+                        Ok(resp) if resp.status().is_success() => return Ok(resp),
+                        Ok(resp) => {
+                            debug!(
+                                "Statsd non-success status code but OK response: {:?}",
+                                resp.status()
+                            );
+                            return Ok(resp);
+                        }
+                        Err(resp) => {
+                            match retry_strategy {
+                                RetryStrategy::LinearBackoff(max_attempts, _)
+                                | RetryStrategy::Immediate(max_attempts)
+                                    if attempts >= max_attempts =>
+                                {
+                                    // handle if status code missing like timeout
+                                    return Err(ShippingError::Destination(
+                                        resp.status(),
+                                        "Failed to send request after {max_attempts}".to_string(),
+                                    ));
+                                }
+                                RetryStrategy::LinearBackoff(_, delay) => {
+                                    tokio::time::sleep(Duration::from_secs(delay)).await;
+                                }
```

Comment:
This will keep re-trying indefinitely until timeout?

Reply:
So this block will just delay and then at the top of the loop (251) we'll increment the attempts. After max attempts is hit, it'll error.

I can add a better test